### ISSUE TITLE
Save image URL locally

### ISSUE THREAD

**Issue by owenfar** [chrome-php/headless-chromium-php#522]
Hello,

I have been trying for the past two days to find a way to download an image URL directly to downloads path or any assigned path without taking a screenshot - to keep origin size & transparency.

I have tried various ways such, one method which I thought would really work is to inject a script that creates an `a` element with href set to the same URL, and a download attribute - which I force click with JS in the same script. This didn't work either though. 

I am using this library to avoid the 403/405 errors that show up because of CloudFlare or similar services. But I cannot find a way or if it is even possible to do this (save URL image as image), maybe someone here knows a hack or if there's an official method I can use?

So basically all I need is to navigate to URL that is an _image URL_, and then once loaded I want to download this image to a set path.

Would be great if someone can help me here.
Thanks in advance

**Comment by owenfar** [comment]
Hey everyone,

I have finally managed to find a way to do this:

```
function downloadImage($img_url, $img_path, $img_ext = "png")
{
    $downloaded = null;

    // Create a new BrowserFactory instance.
    $browserFactory = new BrowserFactory();
    $browserFactory->addOptions(getOptions(true));

    // Create a new Browser instance.
    $browser = $browserFactory->createBrowser();

    try
    {
        // Create a new page in the browser
        $page = $browser->createPage();

        // Navigate to page & wait
        $page->navigate($img_url)->waitForNavigation();
        
        // TEMP - Download to current path for testing
        $img_path = getcwd() . '/';

        // Set the download path
        $page->setDownloadPath($img_path);

        // Set unique image name
        $img_name = time() . '.' . $img_ext;
        
        // Write a script that auto-triggers download
        $script = '
            var xhr = new XMLHttpRequest();
            xhr.open("GET", "' . $img_url . '", true);
            xhr.responseType = "blob";

            xhr.onload = function() {
                if (this.status === 200) {
                    var blob = new Blob([this.response], { type: "image/png" });

                    var link = document.createElement("a");
                    link.href = window.URL.createObjectURL(blob);
                    link.download = "' . $img_name . '";
                    document.body.appendChild(link);

                    link.click();
                }
            };

            xhr.send();
        ';

        // Inject the script
        $page->evaluate($script);

        // Wait for download to complete
        sleep(1);

        // Check if the file exists at the specified path
        if (file_exists($img_path . $img_name))
        {
            // Image downloaded successfully
            $downloaded = $img_name;
        }
    } 
    catch (Exception $e)
    {
        $downloaded = false; // Error occurred during the process
    }
    finally
    {
        $browser->close();
    }

    return $downloaded;
}
```

Navigating to a valid URL was necessary for the injected script to work.

If anyone knows a better way or if there's an "official" way to do this, please let us know.

**EDIT:** Included file_exists check to be 100% sure download was successful
